File: src/main/java/com/sudosoo/takeiteasy/aspect/LoggingAspect.java
package com.sudosoo.takeiteasy.aspect;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Map;

@Aspect
@Component
@RequiredArgsConstructor
public class LoggingAspect {

    private static final Logger logger = LoggerFactory.getLogger(LoggingAspect.class);
    private final ObjectMapper objectMapper;

    // 모든 컨트롤러 && NotLogging 어노테이션 미설정 시 로그 수집
    @Pointcut("within(*..*Controller) && !@annotation(com.sudosoo.takeiteasy.common.NotLogging)")
    public void onRequest() {}

    @Around("onRequest()")
    public Object requestLogging(ProceedingJoinPoint joinPoint) throws Throwable {

        // API 요청 정보
        final RequestApiInfo apiInfo = new RequestApiInfo(joinPoint, joinPoint.getTarget().getClass(), objectMapper);

        // 로그 정보
        final LogInfo logInfo = new LogInfo(
                apiInfo.getUrl(),
                apiInfo.getName(),
                apiInfo.getMethod(),
                apiInfo.getHeader(),
                objectMapper.writeValueAsString(apiInfo.getParameters()),
                objectMapper.writeValueAsString(apiInfo.getBody()),
                apiInfo.getIpAddress()
        );

        try {
            final Object result = joinPoint.proceed(joinPoint.getArgs());

            // Method가 Get이 아닌 로그만 수집
            if (!logInfo.getMethod().equals("GET")) {
                final String logMessage = objectMapper.writeValueAsString(Map.entry("logInfo", logInfo));
                logger.info(logMessage);
            }

            return result;
        } catch (Exception e) {
            final StringWriter sw = new StringWriter();
            e.printStackTrace(new PrintWriter(sw));
            final String exceptionAsString = sw.toString();

            // 발생 Exception 설정
            logInfo.setException(exceptionAsString);
            final String logMessage = objectMapper.writeValueAsString(logInfo);
            logger.error(logMessage);

            throw e;
        }
    }

}

File: takeItEasy-aop/src/main/kotlin/com/sudosoo/takeItEasy/aop/LoggingAspect.kt
package com.sudosoo.takeItEasy.aop

import com.fasterxml.jackson.databind.ObjectMapper
import com.sudosoo.takeItEasy.aop.logging.LogInfo
import com.sudosoo.takeItEasy.aop.logging.RequestApiInfo
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Configuration
import org.springframework.stereotype.Component
import java.io.PrintWriter
import java.io.StringWriter
import kotlin.Any
import kotlin.Exception
import kotlin.Throws
import kotlin.to

@Aspect
@Configuration
class LoggingAspect(private val objectMapper: ObjectMapper) {

    companion object {
        private val logger: Logger = LoggerFactory.getLogger(LoggingAspect::class.java)
    }

    // 모든 컨트롤러 && NotLogging 어노테이션 미설정 시 로그 수집
    @Pointcut("within(*..*Controller) && !@annotation(com.sudosoo.takeItEasy.application.annotation.NotLogging)")
    fun onRequest() {}


    @Around("onRequest()")
    @Throws(Throwable::class)
    fun requestLogging(joinPoint: ProceedingJoinPoint): Any? {

        // API 요청 정보
        val apiInfo = RequestApiInfo(joinPoint, joinPoint.target.javaClass, objectMapper)

        // 로그 정보
        val logInfo = LogInfo(
            apiInfo.url,
            apiInfo.name,
            apiInfo.method,
            apiInfo.header,
            objectMapper.writeValueAsString(apiInfo.parameters).replace("\\", ""),
            objectMapper.writeValueAsString(apiInfo.body).replace("\\", ""),
            apiInfo.ipAddress
        )

        return try {
            val result = joinPoint.proceed(joinPoint.args)

            // Method가 Get이 아닌 로그만 수집
            if (logInfo.method != "GET") {
                val logMessage = objectMapper.writeValueAsString(mapOf("logInfo" to logInfo))
                logger.info(logMessage)
            }
            result

        } catch (e: Exception) {
            val sw = StringWriter()
            e.printStackTrace(PrintWriter(sw))
            val exceptionAsString = sw.toString()

            // 발생 Exception 설정
            logInfo.exception = exceptionAsString
            val logMessage = objectMapper.writeValueAsString(logInfo)
            logger.error(logMessage)

            throw e
        }
    }
}
